### PR TITLE
Olm fallback key support

### PR DESCRIPTION
**Pull request by stefanceriu**
Implements vector-im/element-ios/issues/4406 - Olm fallback key support. NB this PR requires the current development version of Olm (future 3.2.5 release presumably)

### CODE REVIEW

File: MatrixSDK/MXRestClient.h [hunk 1]
```diff
@@ -2122,13 +2122,16 @@ Get the maximum size a media upload can be in bytes.
...
  @param deviceKeys the device keys to send.
  @param oneTimeKeys the one-time keys to send.
+ @param fallbackKeys the fallback keys to send.
 
  @param success A block object called when the operation succeeds.
  @param failure A block object called when the operation fails.
 
  @return a MXHTTPOperation instance.
  */
-- (MXHTTPOperation*)uploadKeys:(NSDictionary*)deviceKeys oneTimeKeys:(NSDictionary*)oneTimeKeys
```
Comment (pixlwave) [inline MatrixSDK/MXRestClient.h:2131]: Should this old method remain for API compatibility?

File: MatrixSDKTests/MatrixSDKTestsData.h [hunk 2]
```diff
@@ -91,7 +91,7 @@ FOUNDATION_EXPORT NSString * const kMXTestsAliceAvatarURL;
 
 
 #pragma mark - mxAlice
-@property (nonatomic, readonly) MXCredentials *aliceCredentials;
+@property (nonatomic, strong, readonly) MXCredentials *aliceCredentials;
```
Comment (pixlwave) [inline MatrixSDKTests/MatrixSDKTestsData.h:94]: This is more a question for me to learn from: I thought `strong` was the default - is this for consistency, or are there conditions where the defaults can be changed?

File: MatrixSDK/MXRestClient.h (hunk 1 above)
Comment (stefanceriu) [reply MatrixSDK/MXRestClient.h:2131]: I thought about it but seeing as we don't have separate methods for uploading keys or oneTimeKeys I figured this is more straight forward.

File: MatrixSDKTests/MatrixSDKTestsData.h (hunk 2 above)
Comment (stefanceriu) [reply MatrixSDKTests/MatrixSDKTestsData.h:94]: Yes, this is just for consistency. strong is indeed the default for reference types and assign for value types.

File: MatrixSDKTests/MXRestClientTests.m [hunk 3]
```diff
@@ -1569,21 +1571,28 @@ - (void)testDeviceKeys
...
-            XCTAssertEqual(keysUploadResponse.oneTimeKeyCounts.count, 1, @"There is no yet one-time keys");
-
-            XCTAssertEqual([keysUploadResponse oneTimeKeyCountsForAlgorithm:@"curve25519"], 2, @"It must response 2 for 'curve25519'");
-            XCTAssertEqual([keysUploadResponse oneTimeKeyCountsForAlgorithm:@"deded"], 0, @"It must response 0 for any other algo");
+            
+            XCTAssertEqual(keysUploadResponse.oneTimeKeyCounts[@"curve25519"].unsignedIntValue, 2, @"Key count must be 2 for 'curve25519'");
+            XCTAssertEqual([keysUploadResponse oneTimeKeyCountsForAlgorithm:@"curve25519"], 2, @"Key count must be 2 for 'curve25519'");
+            
+            [keysUploadResponse.oneTimeKeyCounts enumerateKeysAndObjectsUsingBlock:^(NSString *key, NSNumber *obj, BOOL *stop) {
+                if ([key isEqualToString:@"curve25519"]) {
```
Comment (manuroe) [inline MatrixSDKTests/MXRestClientTests.m:1589]: The previous test `XCTAssertEqual([keysUploadResponse oneTimeKeyCountsForAlgorithm:@"deded"], 0, @"It must response 0 for any other algo");` was more a unit test to validate that `oneTimeKeyCountsForAlgorithm` does not return return a value for a random algo name. We should keep it.

The code of this new block is useless. The server will never return other keys than "curve25519"

File: MatrixSDK/MXRestClient.m [hunk 4]
```diff
@@ -4001,14 +4001,18 @@ - (MXHTTPOperation*)searchUsers:(NSString*)pattern
...
+                  fallbackKeys:(NSDictionary *)fallbackKeys
                        success:(void (^)(MXKeysUploadResponse *keysUploadResponse))success
                        failure:(void (^)(NSError *error))failure
 {
-    return [self uploadKeys:deviceKeys oneTimeKeys:oneTimeKeys forDeviceWithId:nil success:success failure:failure];
+    return [self uploadKeys:deviceKeys oneTimeKeys:oneTimeKeys fallbackKeys:fallbackKeys forDeviceWithId:nil success:success failure:failure];
 }
 
-- (MXHTTPOperation*)uploadKeys:(NSDictionary*)deviceKeys oneTimeKeys:(NSDictionary*)oneTimeKeys
+- (MXHTTPOperation*)uploadKeys:(NSDictionary*)deviceKeys
```
Comment (manuroe) [inline MatrixSDK/MXRestClient.m:4013]: The signature of this method in .h must be updated too. I am surprised there is no build error about it.

File: MatrixSDK/Crypto/MXCrypto.m [hunk 5]
```diff
@@ -2880,6 +2909,35 @@ - (MXHTTPOperation *)generateAndUploadOneTimeKeys:(NSUInteger)keyCount retry:(BO
...
+        signedKey[@"signatures"] = [self signObject:signedKey];
+        
+        fallbackKeyJson[[NSString stringWithFormat:@"%@:%@", kMXKeySignedCurve25519Type, keyId]] = signedKey;
+    }
+    
+    MXLogDebug(@"[MXCrypto] generateAndUploadFallbackKey: Started uploading fallback key.");
+    
+    MXWeakify(self);
+    return [_matrixRestClient uploadKeys:nil oneTimeKeys:nil fallbackKeys:fallbackKeyJson success:^(MXKeysUploadResponse *keysUploadResponse) {
+        weakself.uploadFallbackKeyOperation = nil;
```
Comment (manuroe) [inline MatrixSDK/Crypto/MXCrypto.m:2933]: We never use `weakself` directly but `MXStrongifyAndReturnIfNil(self);` instead.
This method logs if the weak ref has been released. This is useful when analysing rageshake logs.

Comment (stefanceriu) [reply MatrixSDK/Crypto/MXCrypto.m:2933]: Right, got you! 👍

File: MatrixSDKTests/MXRestClientTests.m (hunk 3 above)
Comment (stefanceriu) [reply MatrixSDKTests/MXRestClientTests.m:1589]: 👍

**Review by manuroe — APPROVED** [review]

**Review by manuroe — APPROVED** [review]

**Pull request merged**